DOCS-#2720: Add documentation for partition API (#2721)
modin-project/modin: docs/developer/pandas/partition_api.rst

Partition API in Modin
======================

When you are working with a Modin Dataframe, you can unwrap its remote partitions
to get the raw futures objects compatible with the execution engine (e.g. ``ray.ObjectRef`` for Ray).
In addition to unwrapping of the remote partitions we also provide an API to construct a ``modin.pandas.DataFrame``
from raw futures objects.

Partition IPs
-------------

> Also, you can get the IPs of the nodes that hold the partitions. In that case you can pass

Reviewer: ```suggestion
For finer grained placement control, Modin also provides an API to get the IP addresses of the nodes that hold each partition, ``map_partitions_to_ips``.
```

Partition API in Modin
======================

When you are working with a Modin Dataframe, you can unwrap its remote partitions

> to get the raw future objects compatible with the execution engine (e.g. ``ray.ObjectRef`` for Ray).

Reviewer: ```suggestion
to get the raw futures objects compatible with the execution engine (e.g. ``ray.ObjectRef`` for Ray).
```

> In addition to unwrapping of the remote partitions we also provide API to construct Modin DataFrame from them.

Reviewer: ```suggestion
In addition to unwrapping of the remote partitions we also provide an API to construct
a ``modin.pandas.DataFrame`` from raw futures objects.
```

> API

Reviewer: Make each API a separate section
```suggestion
unwrap_partitions
```

---

It is currently supported the following API:

.. automodule:: modin.distributed.dataframe.pandas
  :noindex:
  :members: unwrap_partitions

.. automodule:: modin.distributed.dataframe.pandas
  :noindex:
  :members: from_partitions

> Running an example with Partition API

Reviewer: ```suggestion
Example
```

-------------------------------------

.. code-block:: python

  import modin.pandas as pd
  from modin.distributed.dataframe.pandas import unwrap_partitions, from_partitions
  import numpy as np
  data = np.random.randint(0, 100, size=(2 ** 10, 2 ** 8))
  df = pd.DataFrame(data)
  partitions = unwrap_partitions(df, axis=0)
  print(partitions)
  # Also, you can create Modin DataFrame from remote partitions
  new_df = from_partitions(partitions, axis=0)
  print(new_df)

Partition IPs Usage
-------------------

> Also, you can use the mentioned above API to get the IPs of the nodes that hold the partitions (``get_ip=True``).

Reviewer: If we are adding a new API to do this, should we wait to update this document?
```suggestion
You can use the mentioned above APIs to get the IPs of the nodes that hold the partitions (``get_ip=True``).
```

Author: Yes, if we add a new API, the doc should be updated. At first we should merge #2643 .

* You can also start Ray as follows: ``ray start --head --system-config='{"max_direct_call_object_size":<nbytes>}'``.

Note that when specifying the threshold the performance of some Modin operations may change.

API
---

It is currently supported the following API:

.. automodule:: modin.distributed.dataframe.pandas

>   :members: unwrap_partitions
> 

Reviewer 2: ```suggestion
  :noindex:
  :members: unwrap_partitions
```

For exclude warning in building process

.. automodule:: modin.distributed.dataframe.pandas

>   :members: from_partitions

Reviewer 2: ```suggestion
  :noindex:
  :members: from_partitions
```

The same

Running an example with pandas partition API
--------------------------------------------

Before you run this, please make sure you follow the instructions listed above.

.. code-block:: python

  import modin.pandas as pd
  from modin.distributed.dataframe.pandas import unwrap_partitions, from_partitions

>   df = pd.read_csv("/path/to/your/file")

Reviewer 2: It might be better to create dataframe from list, for example (in this case user should only copy-paste example code block and see output).

Author: Yep, it makes sense.

>   partitions = unwrap_partitions(df, axis=0, bind_ip=True)

Reviewer 2: ```suggestion
  partitions = unwrap_partitions(df, axis=0, get_ip=True)
```

> Pandas Partition API in Modin

Reviewer: ```suggestion
Partition API in Modin
```

Author: Removed.

=============================

If you are working with a Modin Dataframe, you can unwrap its remote partitions
to get the raw future objects compatible with the execution engine (e.g. ``ray.ObjectRef`` for Ray).

> You can use this API to get the IPs of the nodes that hold these objects as well. In that case you can pass the partitions

Reviewer: Create a separate section to talk about IPs. It is more of an advanced feature, so talk about it later in one place.

Author: Created the separate section for this.

having needed IPs to your function. It can help with minimazing of data movement between nodes. In addition to
unwrapping of the remote partitions we also provide API to construct Modin DataFrame from them.

Ray engine
----------

> However, it is worth noting that for Modin on ``Ray`` engine with ``pandas`` backend IPs of the remote partitions may not match

Reviewer: This paragraph should go in the separate section about IPs.

Author: Moved.

actual locations if the partitions are lower than 100 kB. Ray saves such objects (<= 100 kB, by default) in in-process store
of the calling process (please, refer to `Ray documentation`_ for more information). We can't get IPs for such objects while maintaining good performance.
So, you should keep in mind this for unwrapping of the remote partitions with their IPs. Several options are provided to handle the case in
``How to handle Ray objects that are lower 100 kB`` section.

Dask engine
-----------
There is no mentioned above issue for Modin on ``Dask`` engine with ``pandas`` backend because ``Dask`` saves any objects
in the worker process that processes a function (please, refer to `Dask documentation`_ for more information).

> Install Modin Pandas Partition API

Reviewer: I don't think this installation section is providing any value.

Author: I agree, removed.
